File: pages/oop/2024/labs/03/snippets/01.py
# Базовый класс Train
class Train:
 def __init__(self, speed=0, capacity=0):
    self._speed = speed
    self._capacity = capacity
    def set_speed(self, speed):
        if speed >= 0:
            self._speed = speed
    def get_speed(self):
        return self._speed
    def set_capacity(self, capacity):
        if capacity >= 0:
            self._capacity = capacity
    def get_capacity(self):
        return self._capacity
    def print_info(self):
        print(f"Train with speed: {self._speed} km/h and capacity: {self._capacity} units.")

# Класс для перевозки грузов
class FreightCarrier:
    def load_cargo(self, cargo_weight):
        raise NotImplementedError("This method should be overridden")

# Класс для перевозки пассажиров
class PassengerCarrier:
    def board_passengers(self, passengers):
        raise NotImplementedError("This method should be overridden")

# Класс грузового поезда
class FreightTrain(Train, FreightCarrier):
    def load_cargo(self, cargo_weight):
        print(f"Loaded {cargo_weight} tons of cargo.")
    def print_info(self):
        print(f"Freight Train with speed: {self._speed} km/h and capacity: {self._capacity} tons.")

# Класс пассажирского поезда
class PassengerTrain(Train, PassengerCarrier):
    def board_passengers(self, passengers):
        print(f"Boarded {passengers} passengers.")
    def print_info(self):
        print(f"Passenger Train with speed: {self._speed} km/h and capacity: {self._capacity} passengers.")

# Класс смешанного поезда
class MixedTrain(FreightTrain, PassengerTrain):
    def print_info(self):
        print(f"Mixed Train with speed: {self._speed} km/h, capacity for {self._capacity} tons of cargo and passengers.")

# Тестирование программы
mixed_train = MixedTrain()
mixed_train.set_speed(80)
mixed_train.set_capacity(300)
mixed_train.print_info()
mixed_train.load_cargo(100)
mixed_train.board_passengers(150)

File: pages/oop/2024/labs/03/snippets/01.cpp
#include <iostream>

// Базовый класс Train
class Train {
protected:
    int speed;
    int capacity;
public:
    Train(int spd, int cap) : speed(spd), capacity(cap) {}
    virtual void setSpeed(int spd) {
        speed = spd;
    }
    virtual int getSpeed() const {
        return speed;
    }
    virtual void setCapacity(int cap) {
        capacity = cap;
    }
    virtual int getCapacity() const {
        return capacity;
    }
    virtual ~Train() = default;
};

// Интерфейс для грузоперевозок
class FreightCarrier {
public:
    virtual void loadCargo(int cargo) = 0;
    virtual ~FreightCarrier() = default;
};

// Интерфейс для перевозки пассажиров
class PassengerCarrier {
public:
    virtual void boardPassengers(int passengers) = 0;
    virtual ~PassengerCarrier() = default;
};

// Класс FreightTrain, наследует от Train и реализует FreightCarrier
class FreightTrain : virtual public Train, public FreightCarrier {
public:
    FreightTrain(int spd, int cap) : Train(spd, cap) {}
    void loadCargo(int cargo) override {
        std::cout << "Loading " << cargo << " tons of cargo." << std::endl;
    }
    void printInfo() const {
        std::cout << "Freight Train with speed: " << speed << " km/h, capacity for " << capacity << " tons of cargo." << std::endl;
    }
};

// Класс PassengerTrain, наследует от Train и реализует PassengerCarrier
class PassengerTrain : virtual public Train, public PassengerCarrier {
public:
    PassengerTrain(int spd, int cap) : Train(spd, cap) {}
    void boardPassengers(int passengers) override {
        std::cout << "Boarding " << passengers << " passengers." << std::endl;
    }
    void printInfo() const {
        std::cout << "Passenger Train with speed: " << speed << " km/h, capacity for " << capacity << " passengers." << std::endl;
    }
};

// Класс MixedTrain, наследует как от FreightTrain, так и от PassengerTrain
class MixedTrain : public FreightTrain, public PassengerTrain {
public:
    MixedTrain(int spd, int cap) : Train(spd, cap), FreightTrain(spd, cap), 
    PassengerTrain(spd, cap) {}
    void printInfo() const {
        std::cout << "Mixed Train with speed: " << getSpeed() << " km/h, capacity for " << getCapacity() << " passengers and cargo." << std::endl;
    }
};

// Основная функция
int main() {
    MixedTrain mt(120, 300);
    mt.printInfo();
    mt.loadCargo(50);
    mt.boardPassengers(200);
    return 0;
}
